google/resource_dataproc_cluster.go

```diff
@@ -0,0 +1,870 @@
+	}
+	if t[len(t)-1:] != "s" {
+		return 0, fmt.Errorf("Unexpected init timeout format expecting in seconds e.g. ZZZs, found : %s", t)
+	}
+
+	tsec, err := strconv.Atoi(t[:len(t)-1])
+	if err != nil {
+		return 0, fmt.Errorf("Cannot convert init timeout to int: %s", err)
+	}
+	return tsec, nil
+}
+
+func resourceDataprocClusterDelete(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	project, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+
+	region := d.Get("region").(string)
+	clusterName := d.Get("name").(string)
+	deleteAutoGenBucket := d.Get("delete_autogen_bucket").(bool)
+	timeoutInMinutes := int(d.Timeout(schema.TimeoutDelete).Minutes())
+
+	if deleteAutoGenBucket {
+		if err := deleteAutogenBucketIfExists(d, meta); err != nil {
+			return err
+		}
+	}
+
+	log.Printf("[DEBUG] Deleting Dataproc cluster %s", clusterName)
+	op, err := config.clientDataproc.Projects.Regions.Clusters.Delete(
+		project, region, clusterName).Do()
+	if err != nil {
+		return err
+	}
+
+	// Wait until it's deleted
+	waitErr := dataprocClusterOperationWait(config, op, "deleting Dataproc cluster", timeoutInMinutes, 3)
+	if waitErr != nil {
+		return waitErr
+	}
+
+	log.Printf("[INFO] Dataproc cluster %s has been deleted", d.Id())
+	d.SetId("")
+	return nil
+}
+
+func deleteAutogenBucketIfExists(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	// Determine if the user specified a specific override staging bucket, if so
+	// let it be ...  otherwise GCP will have created one for them somewhere along
+	// the way, we auto delete it as this will be dangling around otherwise
+
+	v, specified := d.GetOk("staging_bucket")
+	if specified {
+		log.Printf("[DEBUG] staging bucket %s (for dataproc cluster) has explicitly been set, leaving it...", v)
+		return nil
+	}
+	bucket := d.Get("bucket").(string)
+
+	log.Printf("[DEBUG] Attempting to delete autogen bucket %s (for dataproc cluster) ...", bucket)
+	return emptyAndDeleteStorageBucket(config, bucket)
+}
+
+func emptyAndDeleteStorageBucket(config *Config, bucket string) error {
+	err := deleteStorageBucketContents(config, bucket)
+	if err != nil {
+		return err
+	}
+
+	err = deleteEmptyBucket(config, bucket)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func deleteEmptyBucket(config *Config, bucket string) error {
+	// remove empty bucket
+	err := resource.Retry(1*time.Minute, func() *resource.RetryError {
+		err := config.clientStorage.Buckets.Delete(bucket).Do()
+		if err == nil {
+			return nil
+		}
+		gerr, ok := err.(*googleapi.Error)
+		if gerr.Code == 404 {
+			// Bucket may be gone already ignore
+			return nil
+		}
+		if ok && gerr.Code == 429 {
+			return resource.RetryableError(gerr)
+		}
+		return resource.NonRetryableError(err)
+	})
+	if err != nil {
+		fmt.Printf("[ERROR] Attempting to delete autogen bucket (for dataproc cluster) if exists: Error deleting bucket %s: %v\n\n", bucket, err)
```

Comment:
Let's use the full word (autogenerated) just for extra clarity. Also I don't think you need to put down "if exists" since at this point we know it exists.